tools/mpremote/mpremote/main.py

```diff
@@ -463,9 +468,21 @@ def main():
             if pyb is None:
                 pyb = do_connect(["auto"])
 
+            if cmd == "soft-reset" and not soft_reset:
+                if pyb.in_raw_repl:
+                    pyb.exit_raw_repl()
+                soft_reset = True
+                continue
+
+            if cmd == "no-soft-reset" and soft_reset:
+                if pyb.in_raw_repl:
+                    pyb.exit_raw_repl()
+                soft_reset = False
```

Comment:
Why is there a `and soft_reset` in the if-logic?  I think it should run this block of code regardless of whether `soft_reset` is true/false.

Also, it doesn't really need to exit the raw repl.  If it's already in raw repl then it's ok to stay that way.  Removing that would then allow this block to go before the `do_connect()` block above, and so can be used before any connection is made to a device (making it behave more like your original `--no-soft-reset` option).

Reply:
> Why is there a `and soft_reset` in the if-logic? I think it should run this block of code regardless of whether `soft_reset` is true/false.

It's because cmd `no-soft-reset` changes how next `enter_raw_repl()` is called and it is not necessary to do it twice. When in raw repl without soft reset already, we may stay there.

> Also, it doesn't really need to exit the raw repl.

I think it is necessary because the if I want to use these commands multiple times in single command chain there is no other way to do it, I must exit the raw repl and enter it again, e.g.

'mpremote **exec** "aaa=42" no-soft-reset _exec_ "aaa=aaa/7" _eval_ "aaa" soft-reset **exec** "pass"`

It will perform the soft reset before **bold** commands and no soft reset before _italic_ commands.